### DISCUSSION TITLE
Is zigpy radio librariew works on Windows system

### DISCUSSION THREAD

**Discussion by pipiche38** [zigpy/zigpy#949] [Q&A] [answered]
the question is in the title ?

**Comment by MattWestb** [comment]
Im normally having 2 docker container with HA core running with ZHA and network connected EZSP radios and is working OK with 10-20 devices for testing but its 110% supported with HA.
The problem i think if like using local comports getting them working OK under windows and mapping them to the container.

Native windows is no go for my but have testing for more the 2 years but i dont like spamming my OS with more "invasive installations".

**Comment by pipiche38** [comment]
pycrypto doesn't exist on Windows, which one do you install ?

**Reply by MattWestb** [reply]
I have not running Zigpy native on windows (for very long time and have reinstalling Windows sens) so i dont knowing witch is working or not.
Sometime i running Bellows native without Zigpy for sniffing with network connected NCP (LIDL / tuya ZBGW or tasmota / ESPHome) and its working OK but normally large problems getting python working OK with all dependencies.

**Reply by puddly** [reply]
I've had no issues installing zigpy and the radio libraries on Windows. I think the problem is that you are not using Python dependency management correctly: https://github.com/zigbeefordomoticz/Domoticz-Zigbee#pre-requisities=

```
sudo pip3 install voluptuous pycrypto aiosqlite crccheck pyusb attrs aiohttp pyserial-asyncio
```

You should be running `pip3 install zigpy` or `pip3 install -e ./path/to/external/zigpy/source`.

**Reply by pipiche38** [reply]
@puddly we are not using pip to install zigpy and others, because 
1) we have developed the current extension based on the zigpy dev branch
2) we want to control what our users use in order to be able to debug properly
3) we want to make sure that they use what we have tested.
4) the plugin runs  under the embedded python framework of Domoticz. Depending on the OS it is run under root or under the user. The python environment cannot be used either.

This is why we are not using such approach, and fully agree with you that with ZHA which is a full native python is much easier, but it is unfortunately to our environment

**Reply by puddly** [reply]
1. Zigpy depends on `pycryptodome`, not `pycrypto`. This is specified in `setup.py` and you would not run into this problem if you let `pip` take care of resolving dependencies.
 2. Similarly, `zigpy-znp` and every other radio library absolutely cannot have `pyserial-asyncio==0.5.0` installed on Windows. Otherwise, it will break. This is again specified in the `setup.cfg`/`setup.py` file of affected radio libraries.

There *must* be a better way for you to install dependencies. Telling your users to manually run a long `sudo pip3 install ...` command of unpinned packages will result in problems like this.

The very least you can do here is install things in your development environment using pip's dependency resolver (`pip install -e /your/zigpy/fork`) and then freeze the list of packages:  `pip freeze > requirements.txt`. Then, `requirements.txt` will contain *exactly* the versions of the packages you're currently using. You will have to edit that file to remove development dependencies but if your users run `pip install -r requirements.txt`, it will install those exact same packages for them.

For example:

```console
$ virtualenv -p 3.9 venv
$ source venv/bin/activate
(venv) $ pip freeze
# no output, there are no installed packages
(venv) $ pip install zigpy zigpy-znp bellows
...
(venv) $ pip freeze
aiohttp==3.8.1
aiosignal==1.2.0
aiosqlite==0.17.0
async-timeout==4.0.2
attrs==21.4.0
bellows==0.29.0
charset-normalizer==2.0.12
click==8.1.2
click-log==0.4.0
coloredlogs==15.0.1
crccheck==1.1
frozenlist==1.3.0
humanfriendly==10.0
idna==3.3
jsonschema==4.4.0
multidict==6.0.2
pure-pcapy3==1.0.1
pycryptodome==3.14.1
pyrsistent==0.18.1
pyserial==3.5
pyserial-asyncio==0.6
typing_extensions==4.1.1
voluptuous==0.13.0
yarl==1.7.2
zigpy==0.44.1
zigpy-znp==0.7.0
```

You can specify each of these dependencies on the command line (`pip install aiohttp==3.8.1 aiosignal==1.2.0 ...`) but it's easier to read the package list from `requirements.txt`.

**Reply by Hedda** [reply]
> 1. Zigpy depends on `pycryptodome`, not `pycrypto`. This is specified in `setup.py` and you would not run into this problem if you let `pip` take care of resolving dependencies.

@pipiche38 FYI, puddly has now replaced `pycryptodome` with `cryptography` for zigpy in PR -> https://github.com/zigpy/zigpy/pull/973

Obviously, this PR is so new that not yet part of a zigpy release -> https://github.com/zigpy/zigpy/releases

Anyway, that "cryptography" library should also work on Windows operating systems:

https://pypi.org/project/cryptography/

https://github.com/pyca/cryptography

https://cryptography.io/en/latest/

**Reply by pipiche38** [reply]
thanks for informing us,  we will wait for the new-radio-API to update. 

PS/ we have an up and running version of Windows working. so no issue to remain on that

**Comment by pipiche38** [comment]
Thanks @puddly I’ll investigate your suggestion and the freeze

Envoyé de mon iPhone

**Comment by pipiche38** [comment, marked as answer]
Just wanted to confirm that we have now the Zigbee for Domoticz running on windows without any issues
thanks for your support
